src/utilities/scripts/windio.py

```diff
@@ -0,0 +1,256 @@
+    ref = ref.removeprefix("#/definitions/")
+    parts = ref.split("/")
+    name = parts[0]
+    node = definitions[name]
+    definition_map = {k: Schema(v) for k, v in node.items()}
+    schema = definition_map[parts[1]]
+    return name, schema
+
+
+def build_type(field: Field, schema: Schema, definition_map: dict, struct_map: dict[str, Struct]) -> None:
+    """
+    Determines the type of a field based on the schema and builds the Struct/Class if necessary.
+    - If the field is an object, a new Struct is created for it
+    - If the field is a string, the type is set to std::string
+    - If the field is a number, the type is set to double
+    - If the field is an integer, the type is set to int
+    - If the field is a boolean, the type is set to bool
+    - If the field is an array, the type is set to the item type followed by []
+
+    Args:
+        field (Field): The field to build
+        schema (Schema): The schema of the field
+        definition_map (dict): The definitions present in the schema
+        struct_map (dict[str, Struct]): The structs that have already been built based on the schema
+
+    Returns:
+        None
+    """
+    # If this field is a reference to a definition we need to get the schema for that definition
+    if schema.reference:
+        if schema.type: # If the field specifies both $ref and type, raise an error
+            raise ValueError(f"{field.name} specifies $ref and type")
+        _, schema = get_ref(schema.reference, definition_map, struct_map)
+
+    # Set the type to object if no type is specified but properties are present
+    if not schema.type and schema.properties:
+        schema.type = 'object'
+
+    # If the schema has multiple types, use the first one (for now)
+    if schema.one_of:
+        schema = schema.one_of[0]
+
+    # Set the type based on the schema type
+    if schema.type == 'object': # If the field is an object, build a Struct for it
+        field.type = field.name
+        build_structs(Struct(field.name, schema.description), schema, definition_map, struct_map)
+    elif schema.type == 'string':
+        field.type = 'std::string'
+    elif schema.type == 'number':
+        field.type = 'double'
+    elif schema.type == 'integer':
+        field.type = 'int'
+    elif schema.type == 'boolean':
+        field.type = 'bool'
+    elif schema.type == 'array':
+        if not schema.items:
+            raise ValueError(f"{field.name}: array without item spec")
+        field.type = schema.items.type
+        build_type(field, schema.items, definition_map, struct_map)
+        field.type = f"std::vector<{field.type}>"
+    else:
+        raise ValueError(f"Unknown type '{schema.type} - {schema}'")
+
+    # print the detected type
+    #print(f"{field.name}: {field.type}")
+
+
+def main():
+    """
+    Main function to build the structs from the schema
+    """
+    with open("/Users/fbhuiyan/dev/openturbine/src/utilities/scripts/IEAontology_schema.yaml", 'r') as file:
```

Comment:
This hardcodes your local path into the script.  Perhaps the input and output file names should be taken as command line data, so running the script should look something like `./windio.py -i schema.yaml -o structs.cpp`

Reply:
Fixed